source/s2n/s2n_tls_channel_handler.c

```diff
@@ -1019,6 +1076,50 @@ static int s_s2n_handler_shutdown(
             s2n_handler->state = NEGOTIATION_FAILED;
         }
 
+        if (!abort_immediately && s2n_handler->state == NEGOTIATION_SUCCEEDED &&
+            !aws_linked_list_empty(&s2n_handler->input_queue) && slot->adj_right) {
+            /**
+             * In case of if we have any queued data in the handler after negotiation and we start to shutdown,
+             * make sure we pass those data down the pipeline before we complete the shutdown.
+             */
+            size_t downstream_size = aws_channel_slot_downstream_read_window(slot);
+
+            AWS_LOGF_DEBUG(
+                AWS_LS_IO_TLS,
+                "id=%p: TLS handler still have pending data to be delivered during shutdown. Wait until downstream "
+                "reads the data.",
+                (void *)handler);
+            if (downstream_size == 0) {
+                AWS_LOGF_WARN(
+                    AWS_LS_IO_TLS,
+                    "id=%p: TLS handler have pending data but cannot delivered to downstream because of flow control. "
+                    "It's possible to result in hanging without any further window update.",
+                    (void *)handler);
+            }
+
+            if (s2n_handler->read_delayed_shutdown_task == NULL) {
+                s2n_handler->read_delayed_shutdown_task =
+                    aws_mem_calloc(handler->alloc, 1, sizeof(struct aws_tls_delayed_shutdown_task));
+                aws_channel_task_init(
+                    &s2n_handler->read_delayed_shutdown_task->task,
+                    s_s2n_read_delayed_shutdown_task,
+                    &s2n_handler->handler,
+                    "s2n_read_delayed_shutdown");
+
+                s2n_handler->read_delayed_shutdown_task->slot = slot;
+                s2n_handler->read_delayed_shutdown_task->error = error_code;
+                /* Not schedule the delay shutdown until the handler reads to block. */
+                if (!s2n_handler->read_task.node.next) {
+                    /* Kick off read, in case data arrives with TLS negotiation. Shutdown stars right after negotiation.
+                     * Nothing will kick off read in that case. */
+                    aws_channel_task_init(
+                        &s2n_handler->read_task, s_run_read, handler, "s2n_channel_handler_read_on_window_increment");
+                    aws_channel_schedule_task_now(slot->channel, &s2n_handler->read_task);
+                }
+            }
+            return AWS_OP_SUCCESS;
+        }
+
         while (!aws_linked_list_empty(&s2n_handler->input_queue)) {
```

Comment:
this has the same memory leak the darwin code currently has, where input_queue might not be cleaned if we delay read-shutdown, but then something goes wrong and we don't process everything in the queue